src/likelihoods/gaussian.jl

```diff
@@ -20,3 +20,23 @@ GaussianLikelihood() = GaussianLikelihood(1e-6)
 (l::GaussianLikelihood)(f::Real) = Normal(f, sqrt(l.σ²))
 
 (l::GaussianLikelihood)(fs::AbstractVector{<:Real}) = MvNormal(fs, sqrt(l.σ²))
+
+"""
+    HeteroscedasticGaussianLikelihood(σ²)
+
+Heteroscedastic Gaussian likelihood. This is to be used if we assume that the 
+uncertainity associated with the data follows a Gaussian distribution where the 
+variance varies according to the input.
+
+```math
+    p(y|[f, σ]) = Normal(y | f, σ)
+```
+On calling, this would return a normal distribution with mean `f` and variance σ.
+"""
+struct HeteroscedasticGaussianLikelihood end
+
+@functor HeteroscedasticGaussianLikelihood
+
+(::HeteroscedasticGaussianLikelihood)(f::AbstractVector{<:Real}) = Normal(f[1], exp(f[2]))
+
+(::HeteroscedasticGaussianLikelihood)(fs::AbstractVector) = MvNormal([f[1]  for f in fs], exp.([f[2]  for f in fs]))
```

Comment:
```suggestion
(::HeteroscedasticGaussianLikelihood)(fs::AbstractVector) = Product(Normal.(first.(fs), exp.(last.(fs))))
```

Reply:
Actually, I think it is better to use `MvNormal` here. In general, one shouldn't use `Product` but `product_distribution` (similar to `truncated` vs `Truncated`) since the latter can be implemented in a more efficient way. Since it forwards to `MvNormal` anyway for a vector of normal distributions (see https://github.com/JuliaStats/Distributions.jl/blob/7eeee622565a8abb6540edae7ce5d42cd7586b5d/src/multivariate/product.jl#L62-L66) I'd suggest to use
```suggestion
(::HeteroscedasticGaussianLikelihood)(fs::AbstractVector) = MvNormal(first.(fs), exp.(last.(fs)))
```